collector/ethtool_linux.go

```diff
@@ -0,0 +1,175 @@
+	receivedRegex    = regexp.MustCompile(`_rx_`)
+	transmittedRegex = regexp.MustCompile(`_tx_`)
+)
+
+type EthtoolStats interface {
+	Stats(string) (map[string]uint64, error)
+}
+
+type ethtoolStats struct {
+}
+
+func (e *ethtoolStats) Stats(intf string) (map[string]uint64, error) {
+	return ethtool.Stats(intf)
+}
+
+type ethtoolCollector struct {
+	fs      sysfs.FS
+	entries map[string]*prometheus.Desc
+	logger  log.Logger
+	stats   EthtoolStats
+}
+
+// makeEthtoolCollector is the internal constructor for EthtoolCollector.
+// This allows NewEthtoolTestCollector to override it's .stats interface
+// for testing.
+func makeEthtoolCollector(logger log.Logger) (*ethtoolCollector, error) {
+	fs, err := sysfs.NewFS(*sysPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open sysfs: %w", err)
+	}
+
+	// Pre-populate some common ethtool metrics.
+	return &ethtoolCollector{
+		fs:    fs,
+		stats: &ethtoolStats{},
+		entries: map[string]*prometheus.Desc{
+			"rx_bytes": prometheus.NewDesc(
+				"node_ethtool_received_bytes_total",
+				"Network interface bytes received",
+				[]string{"device"}, nil,
+			),
+			"rx_dropped": prometheus.NewDesc(
+				"node_ethtool_received_dropped_total",
+				"Number of received frames dropped",
+				[]string{"device"}, nil,
+			),
+			"rx_errors": prometheus.NewDesc(
+				"node_ethtool_received_errors_total",
+				"Number of received frames with errors",
+				[]string{"device"}, nil,
+			),
+			"rx_packets": prometheus.NewDesc(
+				"node_ethtool_received_packets_total",
+				"Network interface packets received",
+				[]string{"device"}, nil,
+			),
+			"tx_bytes": prometheus.NewDesc(
+				"node_ethtool_transmitted_bytes_total",
+				"Network interface bytes sent",
+				[]string{"device"}, nil,
+			),
+			"tx_errors": prometheus.NewDesc(
+				"node_ethtool_transmitted_errors_total",
+				"Number of sent frames with errors",
+				[]string{"device"}, nil,
+			),
+			"tx_packets": prometheus.NewDesc(
+				"node_ethtool_transmitted_packets_total",
+				"Network interface packets sent",
+				[]string{"device"}, nil,
+			),
+		},
+		logger: logger,
+	}, nil
+}
+
+func init() {
+	registerCollector("ethtool", defaultDisabled, NewEthtoolCollector)
+}
+
+// NewEthtoolCollector returns a new Collector exposing ethtool stats.
+func NewEthtoolCollector(logger log.Logger) (Collector, error) {
+	return makeEthtoolCollector(logger)
+}
+
+func (c *ethtoolCollector) Update(ch chan<- prometheus.Metric) error {
+	netClass, err := c.fs.NetClass()
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) || errors.Is(err, os.ErrPermission) {
+			level.Debug(c.logger).Log("msg", "Could not read netclass file", "err", err)
+			return ErrNoData
+		}
+		return fmt.Errorf("could not get net class info: %w", err)
+	}
+
+	if len(netClass) == 0 {
+		return fmt.Errorf("no network devices found")
+	}
+
+	for device := range netClass {
+		var stats map[string]uint64
+		var err error
+
+		stats, err = c.stats.Stats(device)
+		if err != nil {
```

Comment:
Any way to distinguish between different errors and ignore only those that point to ethtool not being supported?

Reply:
In all cases if there's an error there's also no stats. I don't think we'd want to do more than skip the interface and continue on, but what I could do is log the error like this:

		var err error
		stats, err = c.stats.Stats(device)
		if err != nil {
			level.Debug(c.logger).Log("msg", "Ethtool stats error", "err", err, "device", device)
		}

		if len(stats) < 1 {
			// No stats returned; device does not support ethtool stats.
			continue
		}

What do you think?

Reply:
Ideally Stats() would return a [sentinel error](https://dave.cheney.net/2016/04/27/dont-just-check-errors-handle-them-gracefully) which we could use here to distinguish between ethtool not being supports, which we should ignore silently, or an actual error during collection of the stats, which we should handle (log, error counter or maybe even failing the collector).

Reply:
Stats() mostly just passes back what unix.Syscall returns. I think the cases you're talking about is `unix.EOPNOTSUPP` and possibly `unix.EPERM` aka `os.IsPermission(err)`. I'll have it log if the error is not `EOPNOTSUPP`. I don't think an opaque errors_total counter would be that useful, and I definitely don't want to give up on stats entirely because one of the system's network interfaces returned a weird error.

Reply:
Ok I've added logging for various kinds of errors and expanded the test coverage to include these new branches.